include/oneapi/dpl/pstl/zip_view_impl.h

```diff
@@ -0,0 +1,388 @@
+
+template <bool Const, typename... Views>
+    requires all_forward<Const, Views...>
+struct declare_iterator_category<Const, Views...> {
+    using iterator_category = std::input_iterator_tag;
+};
+
+template <std::ranges::input_range... Views>
+    requires ((std::ranges::view<Views> && ... ) && (sizeof...(Views) > 0))
+class zip_view : public std::ranges::view_interface<zip_view<Views...>> {
+public:
+    template <typename... Types>
+    using tuple_type = oneapi::dpl::__internal::tuple<Types...>;
+
+private:
+    template <typename _ReturnAdapter, typename _F, typename _Tuple, std::size_t... _Ip>
+    static decltype(auto)
+    apply_to_tuple_impl(_ReturnAdapter __tr, _F __f, _Tuple& __t, std::index_sequence<_Ip...>)
+    {
+        return __tr(__f(oneapi::dpl::__internal::get_impl<_Ip>()(__t))...);
+    }
+
+    template <typename _ReturnAdapter, typename _F, typename _Tuple>
+    static decltype(auto)
+    apply_to_tuple(_ReturnAdapter __tr, _F __f, _Tuple& __t)
+    {
+        return apply_to_tuple_impl(__tr, __f, __t, std::make_index_sequence<sizeof...(Views)>{});
+    }
+
+    template <typename _F, typename _Tuple>
+    static auto
+    apply_to_tuple(_F __f, _Tuple& __t)
+    {
+        apply_to_tuple_impl([](auto...){}, __f, __t, std::make_index_sequence<sizeof...(Views)>{});
+    }
+
+public:
+    zip_view() = default;
+    constexpr zip_view(Views... views) : views_(views...) {}
+
+    template <bool Const>
+    class iterator : declare_iterator_category<Const> {
+    public:
+        using iterator_concept = std::conditional_t<all_random_access<Const, Views...>,
+                                                    std::random_access_iterator_tag,
+                                                    std::conditional_t<all_bidirectional<Const, Views...>,
+                                                                       std::bidirectional_iterator_tag,
+                                                                       std::conditional_t<all_forward<Const, Views...>,
+                                                                                          std::forward_iterator_tag,
+                                                                                          std::input_iterator_tag>>>;
+
+        using value_type = std::conditional_t<!Const, tuple_type<std::ranges::range_value_t<Views>...>,
+                                                      tuple_type<std::ranges::range_value_t<const Views>...>>;
+
+        using difference_type = std::conditional_t<!Const, std::common_type_t<std::ranges::range_difference_t<Views>...>,
+                                                           std::common_type_t<std::ranges::range_difference_t<const Views>...>>;
+        using return_tuple_type = std::conditional_t<!Const, tuple_type<typename std::iterator_traits<std::ranges::iterator_t<Views>>::reference...>,
+            tuple_type<typename std::iterator_traits<std::ranges::iterator_t<const Views>>::reference...>>;
```

Comment:
Consider reverting the conditionals to improve readability. Something like "if const - work with const, otherwise - work with non-const"